src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceWatch.kt

```diff
@@ -16,92 +16,125 @@ import io.fabric8.kubernetes.client.KubernetesClientException
 import io.fabric8.kubernetes.client.dsl.Watchable
+import org.jboss.tools.intellij.kubernetes.model.resource.ResourceKind
+import java.util.concurrent.BlockingDeque
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.LinkedBlockingDeque
 
 /**
  * A watch that listens for changes on the kubernetes cluster and operates actions on a model accordingly.
  * The model is only visible to this watcher by operations that the former provides (addOperation, removeOperation).
  */
 open class ResourceWatch(
-    private val addOperation: (HasMetadata) -> Unit,
-    private val removeOperation: (HasMetadata) -> Unit
+        private val addOperation: (HasMetadata) -> Unit,
+        private val removeOperation: (HasMetadata) -> Unit,
+        protected val watchOperations: BlockingDeque<Runnable> = LinkedBlockingDeque(),
+        watchOperationsRunner: Runnable = WatchOperationsRunner(watchOperations)
 ) {
-    private var watches: MutableMap<Watchable<Watch, Watcher<HasMetadata>>, Watch?> = mutableMapOf()
-
-    fun watchAll(suppliers: Collection<() -> Watchable<Watch, Watcher<HasMetadata>>?>) {
-        suppliers.forEach {
-            try {
-                doWatch(it)
-            } catch (e: Exception) {
-                logger<ResourceWatch>().warn("Could not watch $it", e)
-            }
-        }
-    }
-
-    fun watch(supplier: () -> Watchable<Watch, Watcher<HasMetadata>>?) {
-        try {
-            doWatch(supplier)
-        } catch(e: Exception){
-            throw ResourceException("Could not watch resource.", e)
-        }
+    companion object {
+        @JvmField val WATCH_OPERATION_ENQUEUED: Watch = Watch {  }
     }
 
-    private fun doWatch(supplier: () -> Watchable<Watch, Watcher<HasMetadata>>?) {
-        val watchable = supplier.invoke() ?: return
-        val watch = watchable.watch(ResourceWatcher(addOperation, removeOperation)) ?: return
-        watches[watchable] = watch
-    }
+    protected open val watches: ConcurrentHashMap<ResourceKind<*>, Watch?> = ConcurrentHashMap()
+    private val executor: ExecutorService = Executors.newWorkStealingPool(10)
+    private val watchOperationsRunner = executor.submit(watchOperationsRunner)
 
-    fun getAll(): List<Watchable<Watch, Watcher<HasMetadata>>> {
-        return watches.keys.toList()
+    open fun watch(kind: ResourceKind<out HasMetadata>, supplier: () -> Watchable<Watch, Watcher<in HasMetadata>>?) {
+        logger<ResourceWatcher>().info("Watching $kind resources.")
+        watches.computeIfAbsent(kind) {
+            val watchOperation = WatchOperation(kind,
+                    supplier.invoke(),
+                    watches,
+                    addOperation,
+                    removeOperation)
+            watchOperations.add(watchOperation) // enqueue watch operation
+            WATCH_OPERATION_ENQUEUED // Marker: watch operation submitted
```

Comment:
`ResourceWatch` executes `<resource>#watch()` in a separate thread (https://github.com/redhat-developer/intellij-kubernetes/pull/91/files/3da332a3a7cc67ada1f6fe14eef095f37d9b2405#diff-c7f975b84df9b95fe3756383d91f051bR118). 
```
   protected class WatchOperation<R: HasMetadata>: Runnable {
        override fun run() {
            ...
            val watch: Watch? = watchable?.watch(ResourceWatcher(addOperation, removeOperation))
            ...
```
This separate thread waits for these operations to show up in a queue and executes them once they are available (https://github.com/redhat-developer/intellij-kubernetes/pull/91/files/3da332a3a7cc67ada1f6fe14eef095f37d9b2405#diff-c7f975b84df9b95fe3756383d91f051bR102).
```
    private class WatchOperationsRunner(private val watchOperations: BlockingDeque<Runnable>) : Runnable {
        override fun run() {
            while (true) {
                val op = watchOperations.take()
                op.run()
            }
        }
    }
```
Thus `ResourceWatch#watch` drops a watch operation to the queue so that the thread can pick it up and execute it.
This means that there's a timely gap btw `ResourceWatch#watch` and the watch being executed. 
In order not to watch a resource kind twice, one has to check if a watch already exists. `ResourceWatch#watch` does this check by looking up the existing watches (https://github.com/redhat-developer/intellij-kubernetes/pull/91/files/3da332a3a7cc67ada1f6fe14eef095f37d9b2405#diff-c7f975b84df9b95fe3756383d91f051bR46).
```
watches.computeIfAbsent(kind) {
    ...
}
```
One cannot rely on the actual watch to exist since there's a timely gap btw `ResourceWatch#watch` and the actual watch. I thus put a marker into the list of watches when `ResourceWatch#watch` is called. This signals that a watch operation is imminent. Once the watch is actually executed by the internal thread, the marker is replaced by the actual watch (https://github.com/redhat-developer/intellij-kubernetes/pull/91/files/3da332a3a7cc67ada1f6fe14eef095f37d9b2405#diff-c7f975b84df9b95fe3756383d91f051bR122):
```
protected class WatchOperation<R: HasMetadata> : Runnable {
        override fun run() {
            ...
            watches[kind] = watch // replace placeholder
```
Makes sense?

Looking at the code I now spot that `ResourceWatch#ignore` and `ResourceWatch#clear` now should also take into account that the watch they're operating on could be a marker

Reply:
Thanks for the explanation Andrè, it was very useful. However, my concern was regarding the marker. By reading the kotlin documentation i understand that a companion object is like static. So if you receive two calls at the same time, this marker will only keep the reference to one watch (because it is shared), or not? :thinking:

Reply:
@lstocchi you're completely right about a companion being a static object. It has no reference to anything though. It's just a marker saying "the operation, that will create the watch is pending" and also "I'll occupy this space that's reserved for a watch of resources kind X until the watch is created and is replacing me". In this sense a static is imho completely fit for its duty. Being a static also allows me to identif is as a marker that informs of a pending operation.
Makes sense?

Reply:
Ah ok, i see, thanks. :+1: